OS X:  Error: no matching constructor for initialization of 'hpx::lcos::local::condition_variable_any'

When compiling the current master in OS X using default CMake configuration, I receive the following error:

/Users/tb/Downloads/hpx/hpx/util/tuple.hpp:73:17: error: no matching constructor for initialization of
      'hpx::lcos::local::condition_variable_any'
              : _value(std::forward<U>(value))

Happens with both system clang and llvm 3.8.1 (Homebrew), and with both Boost 1.59.0 and Boost 1.60.0 (also both Homebrew).

Happens when compiling HPX itself.  

I compile using: `cmake -DCMAKE_CXX_COMPILER=/usr/local/Cellar/llvm/3.8.1/bin/clang++ ..`

(With system clang the well known thread-local storage error appears of course, sorry for the confusion).

Here's the output after make:

[  0%] Building CXX object src/CMakeFiles/hpx.dir/pre_main.cpp.o
In file included from /Users/tb/Downloads/hpx/src/pre_main.cpp:9:
In file included from /Users/tb/Downloads/hpx/hpx/runtime.hpp:20:
In file included from /Users/tb/Downloads/hpx/hpx/runtime_fwd.hpp:14:
In file included from /Users/tb/Downloads/hpx/hpx/runtime/basename_registration.hpp:12:
In file included from /Users/tb/Downloads/hpx/hpx/lcos/future.hpp:12:
In file included from /Users/tb/Downloads/hpx/hpx/lcos/detail/future_data.hpp:20:
In file included from /Users/tb/Downloads/hpx/hpx/util/bind.hpp:19:
In file included from /Users/tb/Downloads/hpx/hpx/util/invoke_fused.hpp:13:
/Users/tb/Downloads/hpx/hpx/util/tuple.hpp:73:17: error: no matching constructor for initialization of
      'hpx::lcos::local::condition_variable_any'
              : _value(std::forward<U>(value))
                ^      ~~~~~~~~~~~~~~~~~~~~~~

/Users/tb/Downloads/hpx/hpx/util/tuple.hpp:213:13: note: in instantiation of function template specialization
      'hpx::util::detail::tuple_member<2, hpx::lcos::local::condition_variable_any,
      void>::tuple_member<hpx::util::detail::tuple_member<2, hpx::lcos::local::condition_variable_any, void>
      >' requested here
            tuple_impl(tuple_impl&&) = default;
            ^
/usr/local/Cellar/llvm/3.8.1/bin/../include/c++/v1/utility:352:37: note: in instantiation of template class
      'std::__1::is_convertible<hpx::util::tuple<bool, unsigned long,
      hpx::lcos::local::condition_variable_any>, hpx::util::tuple<bool, unsigned long,
      hpx::lcos::local::condition_variable_any> >' requested here
                                    is_convertible<_U2, _T2>::value>::type* = 0)
                                    ^
/usr/local/Cellar/llvm/3.8.1/bin/../include/c++/v1/utility:280:30: note: while substituting deduced template
      arguments into function template 'pair' [with _U1 = const hpx::naming::gid_type, _U2 =
      hpx::util::tuple<bool, unsigned long, hpx::lcos::local::condition_variable_any>]
struct _LIBCPP_TYPE_VIS_ONLY pair
                             ^
/usr/local/Cellar/llvm/3.8.1/bin/../include/c++/v1/__tree:579:16: note: in instantiation of template class
      'std::__1::__value_type<hpx::naming::gid_type, hpx::util::tuple<bool, unsigned long,
      hpx::lcos::local::condition_variable_any> >' requested here
    value_type __value_;
               ^
/usr/local/Cellar/llvm/3.8.1/bin/../include/c++/v1/__tree:780:16: note: in instantiation of template class
      'std::__1::__tree_node<std::__1::__value_type<hpx::naming::gid_type, hpx::util::tuple<bool, unsigned
      long, hpx::lcos::local::condition_variable_any> >, void *>' requested here
        return static_cast<__node_pointer>
               ^
/usr/local/Cellar/llvm/3.8.1/bin/../include/c++/v1/__tree:1067:22: note: in instantiation of member function
      'std::__1::__tree<std::__1::__value_type<hpx::naming::gid_type, hpx::util::tuple<bool, unsigned long,
      hpx::lcos::local::condition_variable_any> >, std::__1::__map_value_compare<hpx::naming::gid_type,
      std::__1::__value_type<hpx::naming::gid_type, hpx::util::tuple<bool, unsigned long,
      hpx::lcos::local::condition_variable_any> >, std::__1::less<hpx::naming::gid_type>, true>,
      std::__1::allocator<std::__1::__value_type<hpx::naming::gid_type, hpx::util::tuple<bool, unsigned long,
      hpx::lcos::local::condition_variable_any> > > >::__end_node' requested here
    __begin_node() = __end_node();
                     ^
/usr/local/Cellar/llvm/3.8.1/bin/../include/c++/v1/map:882:11: note: in instantiation of member function
      'std::__1::__tree<std::__1::__value_type<hpx::naming::gid_type, hpx::util::tuple<bool, unsigned long,
      hpx::lcos::local::condition_variable_any> >, std::__1::__map_value_compare<hpx::naming::gid_type,
      std::__1::__value_type<hpx::naming::gid_type, hpx::util::tuple<bool, unsigned long,
      hpx::lcos::local::condition_variable_any> >, std::__1::less<hpx::naming::gid_type>, true>,
      std::__1::allocator<std::__1::__value_type<hpx::naming::gid_type, hpx::util::tuple<bool, unsigned long,
      hpx::lcos::local::condition_variable_any> > > >::__tree' requested here
        : __tree_(__vc(key_compare())) {}
          ^
/Users/tb/Downloads/hpx/hpx/runtime/agas/server/primary_namespace.hpp:269:5: note: in instantiation of member
      function 'std::__1::map<hpx::naming::gid_type, hpx::util::tuple<bool, unsigned long,
      hpx::lcos::local::condition_variable_any>, std::__1::less<hpx::naming::gid_type>,
      std::__1::allocator<std::__1::pair<const hpx::naming::gid_type, hpx::util::tuple<bool, unsigned long,
      hpx::lcos::local::condition_variable_any> > > >::map' requested here
    primary_namespace()
    ^
/Users/tb/Downloads/hpx/hpx/lcos/local/condition_variable.hpp:130:11: note: candidate constructor
      (the implicit copy constructor) not viable: no known conversion from 'hpx::util::detail::tuple_member<2,
      hpx::lcos::local::condition_variable_any, void>' to 'const hpx::lcos::local::condition_variable_any' for
      1st argument
    class condition_variable_any
          ^
/Users/tb/Downloads/hpx/hpx/lcos/local/condition_variable.hpp:130:11: note: candidate constructor
      (the implicit default constructor) not viable: requires 0 arguments, but 1 was provided
1 error generated.
make[2]: *** [src/CMakeFiles/hpx.dir/pre_main.cpp.o] Error 1
make[1]: *** [src/CMakeFiles/hpx.dir/all] Error 2
make: *** [all] Error 2

--- a/hpx/util/tuple.hpp
+++ b/hpx/util/tuple.hpp
@@ -68,7 +68,10 @@ namespace hpx { namespace util
               : _value()
             {}
 
-            template <typename U>
+            template <typename U, typename =
+                typename std::enable_if<
+                    !std::is_same<tuple_member, typename std::decay<U>::type>::value
+                >::type>
             explicit HPX_HOST_DEVICE HPX_CONSTEXPR tuple_member(U&& value)
               : _value(std::forward<U>(value))
             {}
@@ -110,7 +113,10 @@ namespace hpx { namespace util
               : T()
             {}
 
-            template <typename U>
+            template <typename U, typename =
+                typename std::enable_if<
+                    !std::is_same<tuple_member, typename std::decay<U>::type>::value
+                >::type>
             explicit HPX_HOST_DEVICE HPX_CONSTEXPR tuple_member(U&& value)
               : T(std::forward<U>(value))
             {}

